mautic/mautic
# Video form select optimization

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | N
| Automated tests included? | Y
| Related user documentation PR URL | /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

The problem was that an instance with 650 forms was hitting a memory limit when editing a landing page.

```
PHP Error: Allowed memory size of 536870912 bytes exhausted (tried to allocate 299008 bytes) - in file vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php - at line 139 
```

This is the problematic query:
```sql
SELECT
  m0_.is_published AS is_published_0,
  m0_.date_added AS date_added_1,
  m0_.created_by AS created_by_2,
  m0_.created_by_user AS created_by_user_3,
  m0_.date_modified AS date_modified_4,
  m0_.modified_by AS modified_by_5,
  m0_.modified_by_user AS modified_by_user_6,
  m0_.checked_out AS checked_out_7,
  m0_.checked_out_by AS checked_out_by_8,
  m0_.checked_out_by_user AS checked_out_by_user_9,
  m0_.id AS id_10,
  m0_.name AS name_11,
  m0_.description AS description_12,
  m0_.alias AS alias_13,
  m0_.form_attr AS form_attr_14,
  m0_.cached_html AS cached_html_15,
  m0_.post_action AS post_action_16,
  m0_.post_action_property AS post_action_property_17,
  m0_.publish_up AS publish_up_18,
  m0_.publish_down AS publish_down_19,
  m0_.template AS template_20,
  m0_.in_kiosk_mode AS in_kiosk_mode_21,
  m0_.render_style AS render_style_22,
  m0_.form_type AS form_type_23,
  m0_.no_index AS no_index_24,
  m0_.uuid AS uuid_25,
  m0_.category_id AS category_id_26
FROM
  mautic_forms m0_ 
```

With this change we are loading only the ID and label instead of all columns and it will also load only published forms.

This is the new query that is generating the form select box:
```sql
SELECT f.id as value, f.name as label FROM mautic_forms f WHERE f.is_published = 1 ORDER BY f.name ASC
```

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
This is theory, I'm not sure if anyone will want to reproduce the memory limit. I think that checking if it works the same way as before with a few forms will do.

1. Create over 650 of forms
2. Set memory limit to less then 500MB
3. Edit a landing page

- You should see the error mentioned in the description.

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Create some published and unpublished forms.
3. Edit a landing page
4. Add the video slot to the legacy page builder
5. Notice that there are published forms in the select box.

![Screen Shot 2021-07-20 at 14 06 29](https://user-images.githubusercontent.com/1235442/126321098-88827555-0787-48ea-9104-317446b939e3.png)

#### Other areas of Mautic that may be affected by the change:
1. Just the video slot configuration form

[//]: # ( As applicable: )
#### List of areas covered by the unit and/or functional tests:
1. The functional test covers that only published forms will appear in the select box.
